edx_sga/sga.py

```diff
@@ -0,0 +1,530 @@
+        answer = {
+            "sha1": sha1,
+            "filename": upload.file.name,
+            "mimetype": mimetypes.guess_type(upload.file.name)[0],
+        }
+        student_id = self.student_submission_id()
+        submissions_api.create_submission(student_id, answer)
+        path = self._file_storage_path(sha1, upload.file.name)
+        if not default_storage.exists(path):
+            default_storage.save(path, File(upload.file))
+        return Response(json_body=self.student_state())
+
+    @XBlock.handler
+    def staff_upload_annotated(self, request, suffix=''):
+        require(self.is_course_staff())
+        upload = request.params['annotated']
+        module = StudentModule.objects.get(pk=request.params['module_id'])
+        state = json.loads(module.state)
+        state['annotated_sha1'] = sha1 = _get_sha1(upload.file)
+        state['annotated_filename'] = filename = upload.file.name
+        state['annotated_mimetype'] = mimetypes.guess_type(upload.file.name)[0]
+        state['annotated_timestamp'] = _now().strftime(
+            DateTime.DATETIME_FORMAT
+        )
+        path = self._file_storage_path(sha1, filename)
+        if not default_storage.exists(path):
+            default_storage.save(path, File(upload.file))
+        module.state = json.dumps(state)
+        module.save()
+        return Response(json_body=self.staff_grading_data())
+
+    @XBlock.handler
+    def download_assignment(self, request, suffix=''):
+        answer = self.get_submission()['answer']
+        path = self._file_storage_path(answer['sha1'], answer['filename'])
+        return self.download(path, answer['mimetype'], answer['filename'])
+
+    @XBlock.handler
+    def download_annotated(self, request, suffix=''):
+        path = self._file_storage_path(
+            self.annotated_sha1,
+            self.annotated_filename,
+        )
+        return self.download(
+            path,
+            self.annotated_mimetype,
+            self.annotated_filename
+        )
+
+    @XBlock.handler
+    def staff_download(self, request, suffix=''):
+        require(self.is_course_staff())
+        submission = self.get_submission(request.params['student_id'])
+        answer = submission['answer']
+        path = self._file_storage_path(answer['sha1'], answer['filename'])
+        return self.download(path, answer['mimetype'], answer['filename'])
+
+    @XBlock.handler
+    def staff_download_annotated(self, request, suffix=''):
+        require(self.is_course_staff())
+        module = StudentModule.objects.get(pk=request.params['module_id'])
+        state = json.loads(module.state)
+        path = self._file_storage_path(
+            state['annotated_sha1'],
+            state['annotated_filename']
+        )
+        return self.download(
+            path,
+            state['annotated_mimetype'],
+            state['annotated_filename']
+        )
+
+    def download(self, path, mimetype, filename):
```

Comment:
Downloads come through the XBlock as well? Same comment as uploads. If we store things remotely, performing downloads via a link to a third party source will circumvent our servers.